bazel-build.el

```diff
@@ -26,24 +26,61 @@
 
 (defun bazel-build (target)
   "Build a Bazel TARGET."
-  (interactive "sbazel build ")
+  (interactive (list (bazel-build--read-target "bazel build ")))
   (bazel-build--run-bazel-command "build" target))
 
 (defun bazel-run (target)
   "Build and run a Bazel TARGET."
-  (interactive "sbazel run ")
+  (interactive (list (bazel-build--read-target "bazel run ")))
   (bazel-build--run-bazel-command "run" target))
 
 (defun bazel-test (target)
   "Build and run a Bazel test TARGET."
-  (interactive "sbazel test ")
+  (interactive (list (bazel-build--read-target "bazel test ")))
   (bazel-build--run-bazel-command "test" target))
 
 (defun bazel-build--run-bazel-command (command target)
   "Run Bazel tool with given COMMAND, e.g. build or run, on the given TARGET."
   (compile
    (mapconcat #'shell-quote-argument (list "bazel" command target) " ")))
 
+(defun bazel-build--read-target (prompt)
+  "Read a Bazel build target from the minibuffer.  PROMPT is a read-only prompt."
+  (let* ((file-name (buffer-file-name))
+         (workspace-root
+          (or (bazel-build--find-workspace-root file-name)
+              (user-error "Not in a Bazel workspace.  No WORKSPACE file found")))
+         (package-name
+          (or (bazel-build--extract-package-name file-name workspace-root)
+              (user-error "Not in a Bazel package.  No BUILD file found")))
+         (initial-input (concat "//" package-name)))
+    (read-string prompt initial-input)))
+
+(defun bazel-build--find-workspace-root (file-name)
+  "Find the root of the Bazel workspace containing FILE-NAME.
+If FILE-NAME is not in a Bazel workspace, return nil."
+  (locate-dominating-file file-name "WORKSPACE"))
+
+(defun bazel-build--extract-package-name (file-name workspace-root)
+  "Return the nearest Bazel package for FILE-NAME under WORKSPACE-ROOT.
+If FILE-NAME is not in a Bazel package, return nil."
+  (let* ((build-file-directory
+          (cl-some (lambda (build-name)
+                     (locate-dominating-file file-name build-name))
+                   '("BUILD.bazel" "BUILD")))
+         (package-name
+          (cond ((not build-file-directory) nil)
+                ((file-equal-p workspace-root build-file-directory) "")
+                ((file-in-directory-p build-file-directory workspace-root)
+                 (file-relative-name build-file-directory workspace-root))
+                (t nil))))
+    ;; Only return package-name if we can confirm it is the local relative
+    ;; file name of a BUILD file.
+    (and package-name
```

Comment:
Probably you should move these checks to the only `cond` branch where they make sense.

Reply:
Done.